### DISCUSSION TITLE
How to take temperature into consideration in ComputeCrystalPlasticityStress and study the effect of temperature on stress-strain curve

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#17691] [Q&A Modules: Phase field] [answered]
Hi Moose experts/users,

I am looking through the materials module and hope to use ComputeCrystalPlasticityStress class to calculate the response of a FCC crystalline solid. The example file given by Moose is https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html.

With the help of some moose experts or users, I succeed in obtaining the stress-strain curve. But this result is for room temperature. The questions are below.
(1) How to take temperature as a factor and study the effect of temperature on the stress-strain curve.
(2) I am confused about what material is used in the example file (https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read_cp_03.i) and how to define my own material property.
(3) how to obtain an input file with a set of Euler angles like this one https://github.com/idaholab/moose/files/6316215/input_file.txt.

The example files above are provided by @PengWei97. Would you or others like to share any good experiences/tips with me? Thanks for your help in advance.

**Comment by dewenyushu** [comment, marked as answer]
Hi @xchengood 

To answer part your questions:

(1) I am recently working on crystal plasticity code development which involves adding temperature effects. I will be working on a PR in the next week or so and will share with you the link here when the PR is up. 

(2) The example input that you are sharing is using the `FiniteStrainUObasedCP` base class. So the material state variables are actually defined and calculated in the `UserObjects` block. From there, it looks like you are using the Kalidindi model, which is documented in the documentation page that you are referencing. For some detailed descriptions of the constitutive law of the material and default values, I would recommend you to look into the individual user object's documentation page. Also, maybe it is helpful to go through the following paper as well:

Surya R Kalidindi, Curt A Bronkhorst, and Lallit Anand. Crystallographic texture evolution in bulk deformation processing of fcc metals. Journal of the Mechanics and Physics of Solids, 40(3):537–569, 1992.

One thing I want you to be aware of is that we are migrating the `FiniteStrainUObasedCP` base class to `ComputeMultipleCrystalPlasticityStress` (see https://github.com/idaholab/moose/pull/17405), which should be on the devel branch soon. The Kalidindi model is also transformed with the new design, which only includes one material class and should make more sense for you to understand the material model.

(3) The Euler angles seems to be generated as random numbers ranging between 0 and 360, I guess?  The number of rows depends on the read type, i.e., number of elements, blocks, or grains, see https://mooseframework.inl.gov/source/userobjects/ElementPropertyReadFile.html

Hope this is helpful!

@jiangwen84 @sapitts please feel free to add if I miss something

**Reply by xchengood** [reply]
Hi @dewenyushu. Thank you for your information indeed. I would like to ask you how is your working on crystal plasticity code development adding temperature effects. Hope you can finish it soon. In addition, could you please take time to look through and try to run this input file?
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/TempDepCRSS/TempDepCRSS.i

I am not able to get an output file with csv format. what is outlier variable residual norms in terminal below
<img width="528" alt="Screen Shot 2021-05-06 at 11 14 51 AM" src="https://user-images.githubusercontent.com/74226455/117331841-c2998f80-ae5c-11eb-9207-c9c0f30d3f51.png">
I have no idea where goes wrong and how to improve it. Thank you.

**Reply by dewenyushu** [reply]
Hi @xchengood ,

Good to know that you've made some progress on your CP simulation. To answer your questions:

1. An initial implementation for adding temperature effects is close to be merged, see #17735 

2. To output a csv file. You will need to define the desired output variables in the 
```
[Postprocessors]

[]
```
block. There are many `Postprocessor` and `Vectorpostprocessor` types you can choose from, based on the variables that you would like to output. See all the types on the source documentation page [here](https://mooseframework.inl.gov/source/)

**Reply by xchengood** [reply]
Hi @dewenyushu. I am appreciative of your information. Could you tell me which is the specific input file that you add temperature effect in the https://github.com/idaholab/moose/pull/17735.  In addition, could you share good experiences or tips for creating and developing a custom application?  Although I repeatedly look through the example and tutorial as well as the development section, I still have no idea how to create my application, like which syntax I should use, which object class I should utilize. For .c and .h file, what programming rules should I take when writing them? I know c++ is the basis but it looks more complicated.

**Reply by dewenyushu** [reply]
@xchengood 

1. In #17735, I have a test input where I added an AuxVariable to mimic a linearly increasing temperature, please see:

modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/thermal_eigenstrain_test.i

If you would like to solve a thermomechanical problem with crystal plasticity, simply declare a temperature variable and add the heat conduction kernels for the temperature variable in your input. You may refer to some existing inputs, such as:

/Users/yushd/projects/moose/modules/combined/test/tests/power_law_creep/power_law_creep.i

2. I remember MOOSE's training session provides sufficient info about creating an application (see https://mooseframework.inl.gov/). Would you mind being more specific about what kind of application you are creating? what specific difficulties that you are seeing so far?

**Reply by xchengood** [reply]
Hi @dewenyushu. Thank you for giving me good instructions.

1. I am not clear about what is the right way to take your input file that you mentioned in 1 in last reply for my use. I just copy this input file to my moose corresponding folder or download your whole files. For the second way, I try to use 'make -j4' to compile it after it goes into tensor_mechanics, I got the following error.
![image](https://user-images.githubusercontent.com/74226455/118685130-72112300-b7c8-11eb-879e-b7a4009c6185.png)

2. what is 'fth_xx', 'fth_yy' and 'fth_zz' in your input file?

Thank you for your help again.

**Reply by dewenyushu** [reply]
Hi @xchengood 

1. From the error message you posted above, the error from libmesh does not seem to be related to the crystal plasticity codes that we are working on. You may need to update your repo by `conda update --all` and try again 

Nevertheless, copying/downloading the files from the PR does not sound ideal - especially after my recent change of the code to account for multiple thermal eigenstrains. 

- If you are not in a hurry, the easiest for you would be to wait for the PR to be merged, then you can just update your local MOOSE repo based on the next or devel branch.

- If you are comfortable with git, the other approach is to pull my `https://github.com/dewenyushu/moose/tree/add_thermal_eigenstrain_for_CP` branch from my github repo. I am sure there are many ways to do so. Need a bit search and exercise on the git commands.

2. The 'fth_xx', 'fth_yy' and 'fth_zz' are the components in the thermal deformation gradient tensors.

**Comment by ngrilli** [comment]
Dear @xchengood 

To add the effect of temperature, you will need to write your own child material class and add temperature as a coupled variable.
You can find an example in my repository:

https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityThermal.C

This material model includes the thermal eigenstrain to model thermal expansion,
temperature dependence of the critical resolved shear stress,
temperature dependence of the thermal expansion coefficient.

This other material:

https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorCPGrain.C

is for the temperature dependence of the elasticity tensor.

The equations of the crystal plasticity model in MOOSE are reported in section 2.1 of:
https://www.sciencedirect.com/science/article/abs/pii/S0927025618305883

The meaning of the parameters in the input file: 
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/TempDepCRSS/TempDepCRSS.i

it is, for instance:

    flowprops = '1 12 0.001 0.1' # slip rate equations parameters 
    hprops = '1.0 541.5 136.0 200.0 2.5' # hardening properties
    gprops = '1 12 136.0' # initial values of slip system resistances (start_slip_sys, end_slip_sys, value)

0.001 is \dot{\gamma}_0 (equation 4)
0.1 is m (equation 4)
541.5 is h_{s_i s_j} (equation 6)
136.0 is g_{s_j} at time 0 (equation 6)
200.0 is g_{sat} (equation 6)
2.5 is a (equation 6)

Trust this helps.

Best Regards,
Nicolò Grilli

**Reply by xchengood** [reply]
> Dear @ngrilli

> 
> To add the effect of temperature, you will need to write your own child material class and add temperature as a coupled variable.
> You can find an example in my repository:
> 
> https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityThermal.C
> 
> This material model includes the thermal eigenstrain to model thermal expansion,
> temperature dependence of the critical resolved shear stress,
> temperature dependence of the thermal expansion coefficient.
> 
> This other material:
> 
> https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorCPGrain.C
> 
> is for the temperature dependence of the elasticity tensor.
> 
> The equations of the crystal plasticity model in MOOSE are reported in section 2.1 of:
> https://www.sciencedirect.com/science/article/abs/pii/S0927025618305883
> 
> The meaning of the parameters in the input file:
> https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/TempDepCRSS/TempDepCRSS.i
> 
> it is, for instance:
> 
> ```
> flowprops = '1 12 0.001 0.1' # slip rate equations parameters 
> hprops = '1.0 541.5 136.0 200.0 2.5' # hardening properties
> gprops = '1 12 136.0' # initial values of slip system resistances (start_slip_sys, end_slip_sys, value)
> ```
> 
> 0.001 is \dot{\gamma}_0 (equation 4) 0.1 is m (equation 4) 541.5 is h_{s_i s_j} (equation 6)
> 136.0 is g_{s_j} at time 0 (equation 6)
> 200.0 is g_{sat} (equation 6)
> 2.5 is a (equation 6)

First, thanks for your help indeed. I try to understand all your provided. But there is still some confusion in my head. Based on this input file,
1. could you explain more about this one in [Functions] block? I guess this is to set compression force, right? Usually, for uniaxial tension or compression test, the rate of compression/tension like 0.1mm/min is set. So what is the meaning of '0.1*t'?
![image](https://user-images.githubusercontent.com/74226455/116577675-4f30d480-a8d6-11eb-9615-dde16a1ef137.png)

2. In [AuxVariables] or [AuxKernels] block, what is meaning of [./fp_xx], [./fp_yy], [./fp_zz] and [./crysrot11]?
3. In [UserObjects] block, the file you used is called  'euler_ang_test.inp'. How can I get one file like this for my own materials?
4. You did great work in that published paper. I saw the material you studied is β-HMX, right? How do you define the properties of this material in the input file? Usually, I just know we can define one material by something like Young's modulus and Poisson's ratio. I do not see any places regarding these aspects in your input file.

Again, I appreciate your help.
> 
> Trust this helps.
> 
> Best Regards,
> Nicolò Grilli

**Reply by xchengood** [reply]
Hi @ngrilli, the information you provided is much helpful for me to study the effect of temperature on uniaxial tension/compression. But I have some confusion about some points. Could you look at my questions in the last reply and help me figure them out? Thank you.
